benchmark/displace.py

```diff
@@ -0,0 +1,69 @@
+"""
+Eager implementation of displacement generation benchmark
+
+Created on Wed Jun 10 12:44:33 2020
+
+@author: Henry Liu
+"""
+import numpy as np
+import scipy as sc
+import tensorflow as tf
+
+import utils
+
+
+def gen_displace_scipy(N):
+    a = np.diag(np.sqrt(np.arange(1, N, dtype=np.complex64)), k=1)
+    a_dag = np.conjugate(a.transpose())
+    vexpm = np.vectorize(sc.linalg.expm, signature="(n,n)->(n,n)")
+
+    def displace(alphas):
+        alphas = np.expand_dims(np.expand_dims(alphas, -1), -1)
+        return vexpm(alphas * a_dag - np.conj(alphas) * a)
+
+    return displace
+
+
+def gen_displace(N):
+    a = utils.creation(N)
+    a_dag = tf.linalg.adjoint(a)
+
+    def displace(alphas):
+        # wrap for broadcasting [num, 1, 1]
+        alphas = tf.expand_dims(tf.expand_dims(alphas, -1), -1)
+        return tf.linalg.expm(alphas * a_dag - tf.math.conj(alphas) * a)
+
+    return displace
+
+
+def gen_displace_BCH(N):
+    a = tf.cast(utils.creation(N), dtype=tf.complex128)
+    a_dag = tf.linalg.adjoint(a)
+
+    # Convert to pos, mom basis
+    q = a_dag + a
+    p = 1j * (a_dag - a)
+
+    # Diagonalize
+    eig_q, U_q = tf.linalg.eigh(q)
+    eig_p, U_p = tf.linalg.eigh(p)
+    U_q_dag = tf.linalg.adjoint(U_q)
+    U_p_dag = tf.linalg.adjoint(U_p)
+
+    # Commutator
+    comm = tf.linalg.diag_part(q @ p - p @ q)
+
+    def displace(alphas):
+        re_a = tf.expand_dims(tf.cast(tf.math.real(alphas), dtype=tf.complex128), -1)
+        im_a = tf.expand_dims(tf.cast(tf.math.imag(alphas), dtype=tf.complex128), -1)
+
+        # Apply Baker-Campbell-Hausdorff
+        expm_q = tf.linalg.diag(tf.math.exp(1j * im_a * eig_q))
+        expm_p = tf.linalg.diag(tf.math.exp(-1j * re_a * eig_p))
+        expm_c = tf.linalg.diag(tf.math.exp(-0.5 * re_a * im_a * comm))
+
+        return tf.cast(
+            U_q @ expm_q @ U_q_dag @ U_p @ expm_p @ U_p_dag @ expm_c, dtype=tf.complex64
+        )
+
+    return displace
```

Comment:
In the rest of the project we will use translation operators instead of displacement operators. The difference is  `T(alpha) = D(alpha/sqrt(2))` so you might need to add a function `translate(alphas)`. Maybe it's not needed in the benchmarks module, but I just wanted to remind you about this because I am super paranoid about these factors of two and so on :)